### PR TITLE
Anyone should be able to activate an account's vote

### LINKED ISSUE #6467: Anyone should be able to activate an account's vote
### Expected Behavior

Any account should be able to activate a vote.

### Current Behavior

The account needs to activate it's own votes to receive rewards

### PR DESCRIPTION
**Pull request by deepakhb2**
### Description

Currently an accounts pending votes can be activated by the same account. This change will allow to activate any other accounts pending vote. But, the account should belong to group i.e is passed. A new method is added to implement this change and non of the existing methods are modified.

### Other change

Upgraded Election contract to inherit InitializableV2.

### Tested

Added unit tests to cover this change.

### Related issues

- Fixes #6467 

### Backwards compatibility

As this change is made by added a new method and it doesn't affect the existing features.

### CODE REVIEW

**Comment by deepakhb2** [comment]
@nambrot The tests are failing due to ElectionTestInstance interface don't have the new property that was added as part of this change. Any idea where this interface is defined? It looks like this is defined in some other repo.

**Comment by nambrot** [comment]
Hmm that is a good question. It's based upon https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/test/ElectionTest.sol#L9:L9 which should inherit the method and therefore should have it during type generation

**Comment by nambrot** [comment]
Is it passing locally for you?

File: packages/protocol/test/governance/voting/election.ts [hunk 1]
```diff
@@ -621,6 +621,145 @@ contract('Election', (accounts: string[]) => {
...
+                value: new BigNumber(value),
+              },
+            })
+          })
+        })
+      })
+
+      describe('when an epoch boundary has not passed since the pending votes were made', () => {
+        it('should revert', async () => {
+          await assertRevert(election.activate(group))
```
Comment (AlexBHarley) [inline packages/protocol/test/governance/voting/election.ts:751]: Should this be `activateByGroup`?

Comment (AlexBHarley) [inline packages/protocol/test/governance/voting/election.ts]: Should this be `activateByGroup`?

File: packages/protocol/test/governance/voting/election.ts [hunk 2]
```diff
@@ -621,6 +621,145 @@ contract('Election', (accounts: string[]) => {
...
+        })
+      })
+    })
+
+    describe('when the voter does not have pending votes', () => {
+      it('should revert', async () => {
+        await assertRevert(election.activate(group))
```
Comment (AlexBHarley) [inline packages/protocol/test/governance/voting/election.ts]: Should this be `activateByGroup`?

File: packages/protocol/test/governance/voting/election.ts [hunk 3]
```diff
@@ -621,6 +621,145 @@ contract('Election', (accounts: string[]) => {
...
+            event: 'ValidatorGroupVoteActivated',
+            args: {
+              account: voter,
+              group,
+              value: new BigNumber(value),
+            },
+          })
+        })
+
+        describe('when another voter activates votes', () => {
```
Comment (AlexBHarley) [inline packages/protocol/test/governance/voting/election.ts:688]: I'm all for more tests but I feel this logic is already tested by the `#activate.when another voter activates votes` test.

**Review by AlexBHarley — COMMENTED** [review]
Nice! I'm a big fan of this functionality, it will be a great UX improvement for staking services.

File: packages/protocol/test/governance/voting/election.ts (hunk 1 above)
Comment (deepakhb2) [reply packages/protocol/test/governance/voting/election.ts]: Thank you @AlexBHarley

File: packages/protocol/test/governance/voting/election.ts (hunk 3 above)
Comment (deepakhb2) [reply packages/protocol/test/governance/voting/election.ts:688]: This is already tested by activate for negative scenarios. In this case it positive test cases for activateByAccount method.

**Comment by deepakhb2** [comment]
Thank you @nambrot for pointing me to the file how the types are generated and I had made typo while calling the method from test.
I tried running the tests locally and I see more fails with other issues. Also, I am not sure if I used the right command to run the tests locally.

**Comment by nambrot** [comment]
What command are you using to run the test? For the initializable error, consider checking out #7254, you just have to make Election inherit from the new `InitializableV2` library instead of the old one

**Comment by deepakhb2** [comment]
@nambrot I tried running tests using `npm run test` command.

**Comment by deepakhb2** [comment]
@m-chrzan I think ElectionTestContract needs some update. I am not sure if InitializableV2 is creating issues during build. https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/test/governance/voting/lockedgold.ts#L33 please let  me know if I am missing something here. Thank you.

**Comment by m-chrzan** [comment]
> @m-chrzan I think ElectionTestContract needs some update. I am not sure if InitializableV2 is creating issues during build. https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/test/governance/voting/lockedgold.ts#L33 please let me know if I am missing something here. Thank you.

Huh, so this appears to be a slight bug in the test that hadn't been caught before. To fix this particular error message, it should be enough to change that line to

```
const Election: ElectionContract = artifacts.require('Election')
```

(since we're `require`ing the Election contract, not the ElectionTest one) and then propagating those changes throughout the file (changing imports, changing references to `ElectionTest` to `Election`).

I imagine you might then start running into issues with the ElectionTest contract itself, since it inherits from Election (which now inherits from InitializableV2), so also needs to call its parent constructor. Since ElectionTest will always be used in testing, just doing

```
contract ElectionTest is Election(true) {
```

should be enough.

**Comment by deepakhb2** [comment]
> > @m-chrzan I think ElectionTestContract needs some update. I am not sure if InitializableV2 is creating issues during build. https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/test/governance/voting/lockedgold.ts#L33 please let me know if I am missing something here. Thank you.
> 
> Huh, so this appears to be a slight bug in the test that hadn't been caught before. To fix this particular error message, it should be enough to change that line to
> 
> ```
> const Election: ElectionContract = artifacts.require('Election')
> ```
> 
> (since we're `require`ing the Election contract, not the ElectionTest one) and then propagating those changes throughout the file (changing imports, changing references to `ElectionTest` to `Election`).
> 
> I imagine you might then start running into issues with the ElectionTest contract itself, since it inherits from Election (which now inherits from InitializableV2), so also needs to call its parent constructor. Since ElectionTest will always be used in testing, just doing
> 
> ```
> contract ElectionTest is Election(true) {
> ```
> 
> should be enough.

Thank you @m-chrzan for sharing more details around this issue with solution.

**Comment by deepakhb2** [comment]
@nambrot One required 'certora-test' is failing. Is it related to this PR?

**Comment by yorhodes** [comment]
> @nambrot One required 'certora-test' is failing. Is it related to this PR?

@deepakhb2 This is unrelated and there was a fix merged to master today. I have updated your branch.

File: packages/protocol/contracts/governance/Election.sol [hunk 4]
```diff
@@ -270,6 +276,21 @@ contract Election is
...
+  }
+
+  /**
+   * @notice Converts `account`'s pending votes for `group` to active votes.
+   * @param group The validator group to vote for.
+   * @param account The validateor group account's pending votes to active votes
+   * @return True upon success.
+   * @dev Pending votes cannot be activated until an election has been held.
+   */
+  function activateByAccount(address group, address account) external nonReentrant returns (bool) {
```
Comment (yorhodes) [inline packages/protocol/contracts/governance/Election.sol]: super minor nit
```suggestion
  function activateForAccount(address group, address account) external nonReentrant returns (bool) {
```

**Review by yorhodes — APPROVED** [review]
1 nit

**Comment by nambrot** [comment]
@deepakhb2 Looks like there is a linting error?

**Comment by deepakhb2** [comment]
@nambrot for election.ts file the prettier check is getting passed on my local machine. I am not sure if this is with election.ts file.

**Comment by yorhodes** [comment]
> @nambrot for election.ts file the prettier check is getting passed on my local machine. I am not sure if this is with election.ts file.

Patch for the lint fix
https://gist.github.com/yorhodes/1181c8d1cd64426f2a4af529a64ef3ec

Generated from
`yarn prettier -w --config .prettierrc.js packages/protocol/test/governance/voting/election.ts`

**Comment by nambrot** [comment]
@yorhodes @deepakhb2 Can one of y'all apply the patch, and if it passes then I can merge

**Comment by deepakhb2** [comment]
@nambrot I have applied the patch and lint-checks is passed.

**Pull request merged**